### PR TITLE
Adding GeomLoss to the potential solvers

### PR DESCRIPTION
**Pull request by zaccharieramzi**
In order to do so I had to adapt the # content of [`sinkhorn_tensorized`](https://github.com/jeanfeydy/geomloss/blob/main/geomloss/sinkhorn_samples.py).
Indeed, the vanilla `SamplesLoss` does not let one specify the number of steps you want to run or have a fixed regularization (i.e. `eps`) parameter.

Made sure to use a consistent cost function with 1/2 euclidean squared in order to ease comparison between frameworks and have an easy transport matrix construction for geomloss.

### CODE REVIEW

**Comment by zaccharieramzi** [comment]
Screenshot from 2023-02-14 18-19-37
For now GeomLoss is not respecting the constraints for some reason ...

**Comment by zaccharieramzi** [comment]
Screenshot from 2023-02-15 09-03-03
We now have a decreasing violation of constraints, although much slower for geomloss.

**Comment by zaccharieramzi** [comment]
@tomMoral this is ready for review imho

File: solvers/geomloss.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+with safe_import_context() as import_ctx:
+    from geomloss.sinkhorn_divergence import log_weights, sinkhorn_loop
+    from geomloss.sinkhorn_samples import cost_routines, softmin_tensorized
+    import jax.numpy as jnp
+    import numpy as np
+    import ott
+    from ott.geometry import pointcloud
+    from ott.problems.linear import linear_problem
+    from ott.solvers.linear.sinkhorn import SinkhornOutput
+    import torch
```
Comment (tomMoral) [inline solvers/geomloss.py]:
```suggestion
    import jax.numpy as jnp
    import numpy as np
    
    # Import Geomloss which is based on pytorch.
    import torch
    from geomloss.sinkhorn_divergence import log_weights, sinkhorn_loop
    from geomloss.sinkhorn_samples import cost_routines, softmin_tensorized
    
    # Using OTT to get an output compatible with the objective.
    import ott
    from ott.geometry import pointcloud
    from ott.problems.linear import linear_problem
    from ott.solvers.linear.sinkhorn import SinkhornOutput
```

File: solvers/geomloss.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+
+# The benchmark solvers must be named `Solver` and
+# inherit from `BaseSolver` for `benchopt` to work properly.
```
Comment (tomMoral) [inline solvers/geomloss.py]:
```suggestion
```

File: solvers/geomloss.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    # List of parameters for the solver. The benchmark will consider
+    # the cross product for each key in the dictionary.
+    parameters = {
+        'reg': [1e-2, 1e-1],
+    }
+
+    stopping_criterion = SufficientProgressCriterion(patience=50)
+
+    def set_objective(self, x, a, y, b):
+        # Convert problem into jax array with int32 for jitted computations.
```
Comment (tomMoral) [inline solvers/geomloss.py]:
```suggestion
        # Create a ott problem based on `jax` to compute the output of the solver.

File: solvers/geomloss.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            ), a_jax, b_jax,
+        )
+        self.x, self.a, self.y, self.b = [
+            torch.from_numpy(t).float()[None] for t in (x, a, y, b)
+        ]
+        # put all tensors on GPU if available
+        if torch.cuda.is_available():
+            self.x, self.a, self.y, self.b = [
+                t.cuda() for t in (self.x, self.a, self.y, self.b)
+            ]
```
Comment (tomMoral) [inline solvers/geomloss.py]:
```suggestion

        # Store the problem in torch to use GeomLoss.
        # Use the GPU when it is available.
        device = 'cuda' if torch.cuda.is_available() else 'cpu'
        self.x, self.a, self.y, self.b = [
            torch.from_numpy(t).float().to(device=device)[None]
            for t in (x, a, y, b)
        ]
```

Comment (tomMoral) [reply solvers/geomloss.py]: For the GPU, maybe a good idea would be to set a parameter for this? So we can compare both implementations?

File: solvers/geomloss.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        x, y, a, b = self.x, self.y, self.a, self.b
+        # Retrieve the batch size B, the numbers of samples N, M
+        # and the size of the ambient space D:
+        B, N, D = x.shape
+        _, M, _ = y.shape
+        p = 2
+
+        eps = self.reg
+        eps_list = [eps for _ in range(10*n_iter + 1)]
+        cost = None
```
Comment (tomMoral) [inline solvers/geomloss.py]:
```suggestion
        # Geomloss use the eps schedule to specify the number of iterations.
        # We use a fix schedule here.
        eps = self.reg
        eps_list = [eps for _ in range(10*n_iter + 1)]
        cost = None
```

File: solvers/geomloss.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+        # By default, our cost function :math:`C(x_i,y_j)` is a halved,
+        # squared Euclidean distance (p=2) or a simple Euclidean distance
+        # (p=1):
+        if cost is None:
+            cost = cost_routines[p]
```
Comment (tomMoral) [inline solvers/geomloss.py]: Maybe simplify:
```suggestion
        # Select the squared euclidean cost
        cost = cost_routines[2]
```

File: solvers/geomloss.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+        # Compute the relevant cost matrices C(x_i, y_j), C(y_j, x_i), etc.
+        # Note that we "detach" the gradients of the "right-hand sides":
+        # this is coherent with the way we compute our gradients
+        # in the `sinkhorn_loop(...)` routine,
+        # in the `sinkhorn_divergence.py` file.
+        # Please refer to the comments in this file for more details.
+        C_xy = cost(x, y.detach())  # (B,N,M) torch Tensor
+        C_yx = cost(y, x.detach())  # (B,M,N) torch Tensor
```
Comment (tomMoral) [inline solvers/geomloss.py]: Isn't it simply the transpose here? not sure why both are computed.
But probably better to stick the original code so maybe this is overkill to change this...

**Review by tomMoral — APPROVED** [review]
A few comments to add but LGTM! thanks @zaccharieramzi :)

File: solvers/geomloss.py (hunk 7 above)
Comment (zaccharieramzi) [reply solvers/geomloss.py]: I think it is exactly the transpose, and this is done in order to enable specific gradient computations. I can change it

File: solvers/geomloss.py (hunk 4 above)
Comment (zaccharieramzi) [reply solvers/geomloss.py]: Why not, will implement

File: solvers/geomloss.py (hunk 1 above)
Comment (zaccharieramzi) [reply solvers/geomloss.py]: done

File: solvers/geomloss.py (hunk 2 above)
Comment (zaccharieramzi) [reply solvers/geomloss.py]: done

File: solvers/geomloss.py (hunk 3 above)
Comment (zaccharieramzi) [reply solvers/geomloss.py]: done

File: solvers/geomloss.py (hunk 5 above)
Comment (zaccharieramzi) [reply solvers/geomloss.py]: done

File: solvers/geomloss.py (hunk 6 above)
Comment (zaccharieramzi) [reply solvers/geomloss.py]: done

File: solvers/geomloss.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+                cost_fn=ott.geometry.costs.SqPNorm(p=2)
+            ), a_jax, b_jax,
+        )
+
+        # Store the problem in torch to use GeomLoss.
+        # Use the GPU when it is available.
+        device = (
+            'cuda' if torch.cuda.is_available() and self.use_gpu
+            else 'cpu'
+        )
```
Comment (tomMoral) [inline solvers/geomloss.py]: I would say it is better to skip the solver if `use_gpu` is `True` and no GPU support.

Comment (zaccharieramzi) [reply solvers/geomloss.py]: sure

File: solvers/geomloss.py [hunk 9]
```diff
@@ -32,11 +32,16 @@ class Solver(BaseSolver):
...
-        'use_gpu': [True],
+        'use_gpu': [True, False],
     }
 
     stopping_criterion = SufficientProgressCriterion(patience=50)
 
+    def skip(self, **kwargs):
+        # we skip the solver if use_gpu is True and no GPU is available
+        if self.use_gpu and not torch.cuda.is_available():
+            return True, "No GPU available"
```
Comment (tomMoral) [inline solvers/geomloss.py:43]: This will fail if a gpu is available:
```suggestion
            return True, "No GPU available"
        return False, None
```

Yes I know, the API is not great, it would be better to simply return the reason if skipping, else  `None`.

**Comment by tomMoral** [comment]
Merging, thx @zaccharieramzi :)
